src/backend/distributed/shardsplit/shardsplit_decoder.c

```diff
@@ -62,6 +77,65 @@ _PG_output_plugin_init(OutputPluginCallbacks *cb)
 	/* actual pgoutput callback will be called with the appropriate destination shard */
 	pgoutputChangeCB = cb->change_cb;
 	cb->change_cb = split_change_cb;
+	cb->filter_by_origin_cb = replication_origin_filter_cb;
+}
+
+
+/*
+ * replication_origin_filter_cb call back function filters out publication of changes
+ * originated from any other node other than the current node. This is
+ * identified by the "origin_id" of the changes. The origin_id is set to
+ * a non-zero value in the origin node as part of WAL replication.
+ */
+static bool
+replication_origin_filter_cb(LogicalDecodingContext *ctx, RepOriginId origin_id)
+{
+	if (origin_id != InvalidRepOriginId)
+	{
+		return true;
+	}
+	return false;
+}
+
+
+/*
+ * PublishChangesIfCdcSlot checks if the current slot is a CDC slot. If so, it publishes
+ * the changes as the change for the distributed table instead of shard.
+ * If not, it returns false. It also skips the Citus metadata tables.
+ */
+static bool
+PublishChangesIfCdcSlot(LogicalDecodingContext *ctx, ReorderBufferTXN *txn,
+						Relation relation, ReorderBufferChange *change)
+{
+	char *replicationSlotName = ctx->slot->data.name.data;
+
+	/* Check if the replication slot is CITUS_CDC_SLOT*/
+	if (replicationSlotName != NULL &&
+		strncmp(replicationSlotName, CITUS_SHARD_PREFIX_SLOT, strlen(
+					CITUS_SHARD_PREFIX_SLOT)) != 0)
+	{
+		/* Skip publishing changes for system relations in pg_catalog*/
+		if (relation->rd_rel->relnamespace == PG_CATALOG_NAMESPACE)
+		{
+			return true;
+		}
+
+		char *shardRelationName = RelationGetRelationName(relation);
+		uint64 shardId = ExtractShardIdFromTableName(shardRelationName, true);
+		if (shardId != INVALID_SHARD_ID && ShardExists(shardId))
+		{
+			/* try to get the distributed relation id for the shard */
+			Oid distributedRelationId = RelationIdForShard(shardId);
+			if (OidIsValid(distributedRelationId))
+			{
+				relation = RelationIdGetRelation(distributedRelationId);
```

Comment:
we probably need to rebuild the tuples like we do for split, see also:
https://github.com/citusdata/citus/issues/6303

@saawasek at the time:
>  I tried to further debug this. So what exactly is happening is the source table on which we received concurrent write, has 5 attributes (natts, where one is dropped). The target relation where the tuple should be routed from split decoder has 4 natts. The target relations are created after ALTER TABLE...DROP COLUMN(as we do create_distributed_table_concurrently after alter) hence, the targets are having 4 attributes.
> 
> Now what happens is, the split decoder emits the target relation with 4 attributes which is expected. But the tuple has 5 attributes.
> 
> So now the processing at higher level in pgoutput.c at logicalrep_write_tuple happens with relation descriptor of 4 attributes but tuple having 5 attributes.
> 
> I quickly rewrote the tuple based on the target relation desc and seems it works as expected. I used methods like heap_deform_tuple(to deform tuple)->skip dropped columns->heap_form_tuple.